src/braket/circuits/noise_model/noise_model.py

```diff
@@ -0,0 +1,396 @@
+            set will be populated with the list of targets in the new circuit.
+        """
+        new_circuit = Circuit()
+        for circuit_instruction in circuit.instructions:
+            new_circuit.add_instruction(circuit_instruction)
+            target_qubits = list(circuit_instruction.target)
+            for item in gate_noise_instructions:
+                if item.criteria.instruction_matches(circuit_instruction):
+                    if item.noise.fixed_qubit_count() == len(target_qubits):
+                        new_circuit.add_instruction(Instruction(item.noise, target_qubits))
+                    else:
+                        for qubit in target_qubits:
+                            new_circuit.add_instruction(Instruction(item.noise, qubit))
+        for result_type in circuit.result_types:
+            new_circuit.add_result_type(result_type)
+        return new_circuit
+
+    @classmethod
+    def _apply_init_noise(
+        cls,
+        circuit: Circuit,
+        init_noise_instructions: List[NoiseModelInstruction],
+    ) -> Circuit:
+        """
+        Applies the initialization noise of this noise model to a circuit and returns the circuit.
+
+        Args:
+            circuit (Circuit): A circuit to apply `noise` to.
+            init_noise_instructions (List[NoiseModelInstruction]): A list of initialization noise
+                model instructions.
+
+        Returns:
+            Circuit: The passed in circuit, with the initialization noise applied.
+        """
+        if not init_noise_instructions:
+            return circuit
+        for item in init_noise_instructions:
+            qubits = item.criteria.qubit_intersection(circuit.qubits)
+            if len(qubits) > 0:
+                circuit.apply_initialization_noise(item.noise, list(qubits))
+        return circuit
+
+    @classmethod
+    def _apply_readout_noise(
+        cls,
+        circuit: Circuit,
+        readout_noise_instructions: List[NoiseModelInstruction],
+    ) -> Circuit:
+        """
+        Applies the readout noise of this noise model to a circuit and returns the circuit.
+
+        Args:
+            circuit (Circuit): A circuit to apply `noise` to.
+            readout_noise_instructions (List[NoiseModelInstruction]): The list of readout noise
+                to apply.
+
+        Returns:
+            Circuit: The passed in circuit, with the readout noise applied.
+        """
+        if not readout_noise_instructions:
+            return circuit
+
+        result_types = circuit.result_types
+        if result_types:
+            for result_type in result_types:
+                target_qubits = list(result_type.target)
+                for item in readout_noise_instructions:
+                    if item.criteria.result_type_matches(result_type):
+                        circuit.apply_readout_noise(noise=item.noise, target_qubits=target_qubits)
```

Comment:
Summoning @mbeach-aws and @licedric for help here.

Reply:
@mbeach-aws Can you take a look at the updated [tests](https://github.com/aws/amazon-braket-sdk-python/pull/313/files#diff-925e67ef63651dabc9c31461b9ef409893e711a72b867e2dfeed2bcc7ff5b7eeR428) for the noise model? It may not open up automatically because the diff is large. The unit tests show how I now handle multiple result types and when I use the default.

Reply:
I've now updated the code to not have any assumptions about what is being measured when there are no default result types in the circuit. This means that the customers will need to add observables to the circuit manually, if they want to have the noise model apply readout noise.